src/object/schedcontext.c

```diff
@@ -185,51 +197,61 @@ static exception_t invokeSchedContext_YieldTo(sched_context_t *sc, word_t *buffe
      * if the thread isSchedulable, it is ready and sufficient.*/
     schedContext_resume(sc);
 
-    bool_t return_now = true;
-    if (isSchedulable(sc->scTcb)) {
+    tcb_t *tcb = sc->scTcb;
+
+    bool_t return_now;
+    if (isSchedulable(tcb)) {
         if (SMP_COND_STATEMENT(sc->scCore != getCurrentCPUIndex() ||)
-            sc->scTcb->tcbPriority < NODE_STATE(ksCurThread)->tcbPriority) {
-            tcbSchedDequeue(sc->scTcb);
-            SCHED_ENQUEUE(sc->scTcb);
+            tcb->tcbPriority < NODE_STATE(ksCurThread)->tcbPriority) {
+            tcbSchedDequeue(tcb);
+            SCHED_ENQUEUE(tcb);
+            return_now = true;
         } else {
             NODE_STATE(ksCurThread)->tcbYieldTo = sc;
             sc->scYieldFrom = NODE_STATE(ksCurThread);
-            tcbSchedDequeue(sc->scTcb);
+            tcbSchedDequeue(tcb);
             tcbSchedEnqueue(NODE_STATE(ksCurThread));
-            tcbSchedEnqueue(sc->scTcb);
+            tcbSchedEnqueue(tcb);
             rescheduleRequired();
 
             /* we are scheduling the thread associated with sc,
              * so we don't need to write to the ipc buffer
              * until the caller is scheduled again */
             return_now = false;
         }
+    } else {
+        return_now = true;
     }
 
     if (return_now) {
-        setConsumed(sc, buffer);
+        setConsumed(sc, NODE_STATE(ksCurThread), call);
+        /* Only set to Running if there is a kernel reply message for the user.
+           Restart will create a default empty success message. */
+        setThreadState(NODE_STATE(ksCurThread), ThreadState_Running);
     }
 
     return EXCEPTION_NONE;
 }
 
-static exception_t decodeSchedContext_YieldTo(sched_context_t *sc, word_t *buffer)
+static exception_t decodeSchedContext_YieldTo(sched_context_t *sc, bool_t call)
 {
     if (sc->scTcb == NULL) {
         userError("SchedContext_YieldTo: cannot yield to an inactive sched context");
         current_syscall_error.type = seL4_IllegalOperation;
         return EXCEPTION_SYSCALL_ERROR;
     }
 
-    if (sc->scTcb == NODE_STATE(ksCurThread)) {
+    tcb_t *tcb = sc->scTcb;
+
+    if (tcb == NODE_STATE(ksCurThread)) {
         userError("SchedContext_YieldTo: cannot seL4_SchedContext_YieldTo on self");
         current_syscall_error.type = seL4_IllegalOperation;
         return EXCEPTION_SYSCALL_ERROR;
     }
 
-    if (sc->scTcb->tcbPriority > NODE_STATE(ksCurThread)->tcbMCP) {
+    if (tcb->tcbPriority > NODE_STATE(ksCurThread)->tcbMCP) {
         userError("SchedContext_YieldTo: insufficient mcp (%lu) to yield to a thread with prio (%lu)",
-                  (unsigned long) NODE_STATE(ksCurThread)->tcbMCP, (unsigned long) sc->scTcb->tcbPriority);
+                  (unsigned long) NODE_STATE(ksCurThread)->tcbMCP, (unsigned long) tcb->tcbPriority);
```

Comment:
While you're changing this, you can remove the casts too probably. All platforms have prio_t as an unsigned long.